### PR TITLE
ref #631: fix cronjob scheduler

### PR DESCRIPTION
**Pull request by rnuernberg**
| Q             | A
| ------------- | ---
| Branch        | master for features / 6.3.x
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed issues  | chameleon-system/chameleon-system#631
| License       | MIT

### CODE REVIEW

File: src/CoreBundle/CronJob/CronJobScheduler.php [hunk 1]
```diff
@@ -86,7 +86,7 @@ public function calculateCurrentPlannedExecutionDate(CronJobScheduleDataModel $s
 
         $timePassedSinceLastPlannedExecution = $now->diff($lastPlannedExecution);
 
-        if ($timePassedSinceLastPlannedExecution < $executionInterval) {
```
Comment (UlrichKu) [inline src/CoreBundle/CronJob/CronJobScheduler.php:89]: Did this work before?

Or in what circumstances does this work.

File: src/CoreBundle/CronJob/CronJobScheduler.php [hunk 2]
```diff
@@ -103,6 +103,17 @@ public function calculateCurrentPlannedExecutionDate(CronJobScheduleDataModel $s
...
 
+    private function isLess(\DateInterval $a, \DateInterval  $b): bool
+    {
+        $now = $this->timeProvider->getDateTime();
+        $aNow = clone $now;
+        $aNow = $aNow->add($a);
+        $bNow = clone $now;
+        $bNow->add($b);
+
+        return ($aNow >= $bNow);
```
Comment (UlrichKu) [inline src/CoreBundle/CronJob/CronJobScheduler.php:114]: Shouldn't it be < ?

File: src/CoreBundle/CronJob/CronJobScheduler.php [hunk 3]
```diff
@@ -103,6 +103,17 @@ public function calculateCurrentPlannedExecutionDate(CronJobScheduleDataModel $s
         return $plannedExecution;
     }
 
+    private function isLess(\DateInterval $a, \DateInterval  $b): bool
+    {
+        $now = $this->timeProvider->getDateTime();
```
Comment (UlrichKu) [inline src/CoreBundle/CronJob/CronJobScheduler.php:108]: This seems rather magical/workaround-ish: if necessary a short comment on why would be nice.

File: src/CoreBundle/CronJob/CronJobScheduler.php (hunk 1 above)
Comment (rnuernberg) [reply src/CoreBundle/CronJob/CronJobScheduler.php:89]: removed the check and changed the do...while into a while. This has the same affect and removes complexity.

File: src/CoreBundle/CronJob/CronJobScheduler.php (hunk 2 above)
Comment (rnuernberg) [reply src/CoreBundle/CronJob/CronJobScheduler.php]: the `<` was correct, but the entire method has been removed anyway

File: src/CoreBundle/CronJob/CronJobScheduler.php (hunk 3 above)
Comment (rnuernberg) [reply src/CoreBundle/CronJob/CronJobScheduler.php]: method has been removed

File: src/CoreBundle/private/modules/CMSRunCrons/CMSRunCrons.class.php [hunk 4]
```diff
@@ -71,23 +72,36 @@ public function &Execute()
...
+                   WHERE `start_execution` <= :now
+                     AND (`end_execution` >= :now || `end_execution` = '0000-00-00')
                      AND `active` = '1'
                 ORDER BY `execute_every_n_minutes`
                  ";
+            $cronJobRows = $this->getDatabaseConnection()->fetchAll($sQuery, ['now' => $now]);
+            foreach ($cronJobRows as $cronJobRow) {
+                $cronJobDb = TdbCmsCronjobs::GetNewInstance();
+                // cronjobs may run for a long time - in the meantime other jobs could have been executed by another thread. So load
+                // each job from db just before execution.
```
Comment (UlrichKu) [inline src/CoreBundle/private/modules/CMSRunCrons/CMSRunCrons.class.php:85]: "$now" could then be rather outdated when a previous job took long

File: src/CoreBundle/private/modules/CMSRunCrons/CMSRunCrons.class.php [hunk 5]
```diff
@@ -71,23 +72,36 @@ public function &Execute()
...
-                     AND (`end_execution` >= '".MySqlLegacySupport::getInstance()->real_escape_string($now)."' || `end_execution` = '0000-00-00')
+            $sQuery = "SELECT `id` FROM `cms_cronjobs`
+                   WHERE `start_execution` <= :now
+                     AND (`end_execution` >= :now || `end_execution` = '0000-00-00')
                      AND `active` = '1'
                 ORDER BY `execute_every_n_minutes`
                  ";
+            $cronJobRows = $this->getDatabaseConnection()->fetchAll($sQuery, ['now' => $now]);
+            foreach ($cronJobRows as $cronJobRow) {
+                $cronJobDb = TdbCmsCronjobs::GetNewInstance();
```
Comment (UlrichKu) [inline src/CoreBundle/private/modules/CMSRunCrons/CMSRunCrons.class.php:83]: -> tdbCronJob

(as it is no database here)

Comment (rnuernberg) [reply src/CoreBundle/private/modules/CMSRunCrons/CMSRunCrons.class.php]: renamed

File: src/CoreBundle/private/modules/CMSRunCrons/CMSRunCrons.class.php (hunk 4 above)
Comment (rnuernberg) [reply src/CoreBundle/private/modules/CMSRunCrons/CMSRunCrons.class.php:85]: true

**Pull request merged**